### PR TITLE
enhancement: add chain validation

### PR DESCRIPTION
**Pull request by MarkNerdi**
## Summary

> Please summarize your changes, describing **what** they are and **why** they were made.

...

## Changelog

```
- Please write a list of changes
```

### Platforms

> Please select any platforms where your changes have been tested.

-   **Desktop**
    -   [ ] MacOS
    -   [ ] Linux
    -   [ ] Windows
-   **Mobile**
    -   [ ] iOS
    -   [ ] Android

### Instructions

> Please describe the specific instructions, configurations, and/or test cases necessary to **test and verify** that your changes work as intended.

...

## Checklist

> Please tick the following boxes that are relevant to your changes.

-   [ ] I have followed the contribution guidelines for this project
-   [ ] I have performed a self-review of my own code
-   [ ] I have commented my code, particularly in hard-to-understand areas
-   [ ] I have added or modified tests that prove my changes work as intended
-   [ ] I have verified that new and existing unit tests pass locally with my changes
-   [ ] I have verified that my latest changes pass CI workflows for testing and linting
-   [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: packages/desktop/components/drawers/network-config/drawers/AddChainDrawer.svelte [hunk 1]
```diff
@@ -17,28 +23,53 @@
         iscpEndpoint: '',
     }
 
+    function validate(): boolean {
+        if (!chain.name) {
+            nameError = localize(`${localeKey}.errors.cannotBeEmpty`)
+        }
+
+        const isHex = chain.aliasAddress.startsWith('0x')
+        if (!isHex || !/^(0x08)?[0-9a-f]{64}?(?:0[1-9]|[1-5][0-9]|6[0-4])?0{8}$/i.test(chain.aliasAddress)) {
```
Comment (nicole-obrien) [inline packages/desktop/components/drawers/network-config/drawers/AddChainDrawer.svelte:32]: Can we make a util for this?

**Review by nicole-obrien — DISMISSED** [review]

File: packages/desktop/components/drawers/network-config/drawers/AddChainDrawer.svelte [hunk 2]
```diff
@@ -17,28 +23,53 @@
...
+
+        if (chain.explorerUrl && !isValidUrl(chain.explorerUrl)) {
+            explorerUrlError = localize(`${localeKey}.errors.invalidUrl`)
+        }
+
+        return !!nameError || !!aliasAddressError || !!iscpEndpointError || !!explorerUrlError
+    }
+
     function handleSubmit(): void {
+        validate()
```
Comment (maxwellmattryan) [inline packages/desktop/components/drawers/network-config/drawers/AddChainDrawer.svelte:48]: Probably should just make this function `void` and get rid of a few lines

File: packages/desktop/components/drawers/network-config/drawers/AddChainDrawer.svelte [hunk 3]
```diff
@@ -17,28 +23,53 @@
...
         // TODO: Fetch chainId from ISCP node before adding it to profile
     }
 </script>
 
 <add-chain-drawer class="h-full flex flex-col justify-between">
     <form id="add-chain-form" class="flex flex-col gap-3" on:submit|preventDefault={handleSubmit}>
```
Comment (maxwellmattryan) [inline packages/desktop/components/drawers/network-config/drawers/AddChainDrawer.svelte:54]: Let's go ahead and make the `AddIscpChainForm` component and put the validation logic in there - will help with the upcoming tasks around this `AddChainDrawer` :+1:

File: packages/desktop/components/drawers/network-config/drawers/AddChainDrawer.svelte [hunk 4]
```diff
@@ -17,28 +23,53 @@
...
+        }
+
+        if (chain.explorerUrl && !isValidUrl(chain.explorerUrl)) {
+            explorerUrlError = localize(`${localeKey}.errors.invalidUrl`)
+        }
+
+        return !!nameError || !!aliasAddressError || !!iscpEndpointError || !!explorerUrlError
+    }
+
     function handleSubmit(): void {
```
Comment (maxwellmattryan) [inline packages/desktop/components/drawers/network-config/drawers/AddChainDrawer.svelte:47]: Should be `onSubmitClick`

**Review by maxwellmattryan — CHANGES_REQUESTED** [review]
Have some requested changes :pray:

File: packages/desktop/components/drawers/network-config/drawers/components/AddIscpChainForm.svelte [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        name: '',
+        explorerUrl: undefined,
+        aliasAddress: '',
+        iscpEndpoint: '',
+    }
+
+    function validateName(): void {
+        if (!chain.name) {
+            nameError = localize(`${localeKey}.errors.cannotBeEmpty`)
+        } else if (chain.name.length > 64) {
```
Comment (maxwellmattryan) [inline packages/desktop/components/drawers/network-config/drawers/components/AddIscpChainForm.svelte:29]: Let's make this a constant i.e. `MAX_CHAIN_NAME_LENGTH`

File: packages/desktop/components/drawers/network-config/drawers/components/AddIscpChainForm.svelte [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    function resetErrors(): void {
+        nameError = ''
+        aliasAddressError = ''
+        iscpEndpointError = ''
+        explorerUrlError = ''
+    }
+
+    function onSubmit(): void {
```
Comment (maxwellmattryan) [inline packages/desktop/components/drawers/network-config/drawers/components/AddIscpChainForm.svelte:66]: Should be `onSubmitClick`

File: packages/shared/lib/core/utils/crypto/utils/isValidHexAddress.ts [hunk 7]
```diff
@@ -0,0 +1,4 @@
+export function isValidHexAddress(text: string): boolean {
+    const isHex = text.startsWith('0x')
+    return isHex && !/^(0x08)?[0-9a-f]{64}?(?:0[1-9]|[1-5][0-9]|6[0-4])?0{8}$/i.test(text)
```
Comment (maxwellmattryan) [inline packages/shared/lib/core/utils/crypto/utils/isValidHexAddress.ts:3]: Might be good to extract the regexp into a const, i.e.
```typescript
export const HEX_REGEXP = /^(0x08)?[0-9a-f]{64}?(?:0[1-9]|[1-5][0-9]|6[0-4])?0{8}$/i
```

Comment (maxwellmattryan) [reply packages/shared/lib/core/utils/crypto/utils/isValidHexAddress.ts:3]: We just need to be careful about the `/i` part at the end if we do make it a constant b/c the state is stored in the regexp so as long as we test wrong inputs multiple times back to back, should be alright

File: packages/desktop/components/drawers/network-config/drawers/components/AddIscpChainForm.svelte [hunk 8]
```diff
@@ -0,0 +1,107 @@
+<script lang="ts">
+    import { localize } from '@core/i18n'
+    import { IIscpChainMetadata } from '@core/network'
+    import { ChainType } from '@core/network/enums'
+    import { isValidHexAddress, isValidUrl } from '@core/utils'
+    import { Button, HTMLButtonType, Input } from '@ui'
+
+    const isBusy = false
```
Comment (maxwellmattryan) [inline packages/desktop/components/drawers/network-config/drawers/components/AddIscpChainForm.svelte:8]: Should be `let` but irrelevant at this point

**Review by maxwellmattryan — CHANGES_REQUESTED** [review]
Some minor changes :pray:

File: packages/desktop/components/drawers/network-config/drawers/components/AddIscpChainForm.svelte [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+        if (chain.explorerUrl && !isValidUrl(chain.explorerUrl)) {
+            explorerUrlError = localize(`${localeKey}.errors.invalidUrl`)
+        }
+    }
+
+    function validate(): void {
+        validateName()
+        validateAliasAddress()
+        validateIscpEndpoint()
+        validateExplorerUrl()
```
Comment (maxwellmattryan) [inline packages/desktop/components/drawers/network-config/drawers/components/AddIscpChainForm.svelte:56]: I think part of the validation should include if either the name or alias address is already being used. The explorer URL is fine and the ISCP endpoint is fine so long as the alias address is different.

File: packages/desktop/components/drawers/network-config/drawers/components/AddIscpChainForm.svelte [hunk 10]
```diff
@@ -0,0 +1,107 @@
+<script lang="ts">
+    import { localize } from '@core/i18n'
+    import { IIscpChainMetadata, MAX_CHAIN_NAME_LENGTH } from '@core/network'
+    import { ChainType } from '@core/network/enums'
```
Comment (maxwellmattryan) [inline packages/desktop/components/drawers/network-config/drawers/components/AddIscpChainForm.svelte:4]: Should be able to import this with the other stuff from `@core/network`

**Review by maxwellmattryan — CHANGES_REQUESTED** [review]
Two minor changes :pray:

File: packages/desktop/components/drawers/network-config/drawers/components/AddIscpChainForm.svelte (hunk 9 above)
Comment (maxwellmattryan) [reply packages/desktop/components/drawers/network-config/drawers/components/AddIscpChainForm.svelte:56]: Might be difficult to test that logic but the logic itself should be straightforward, might change slightly once we add the classes in :sweat_smile:

File: packages/shared/locales/en.json [hunk 11]
```diff
@@ -495,7 +495,15 @@
                         "chainId": "Chain ID",
                         "aliasAddress": "Alias Address",
                         "iscpEndpoint": "ISCP Endpoint",
-                        "explorerEndpoint": "Explorer Endpoint (optional)"
+                        "explorerEndpoint": "Explorer Endpoint (optional)",
+                        "errors": {
+                            "cannotBeEmpty": "Name cannot be empty",
+                            "nameTooLong": "Name is too long",
+                            "aliasAddressMustBeHex": "Alias address must be in hex format",
```
Comment (maxwellmattryan) [inline packages/shared/locales/en.json:502]: Is this true? I thought it was Bech32 encoding?

**Review by maxwellmattryan — COMMENTED** [review]
One question :thinking:

**Review by maxwellmattryan — APPROVED** [review]
LGTM :fire:

**Pull request merged**